uvicorn/protocols/websocket.py

```diff
@@ -0,0 +1,149 @@
+import asyncio
+import time
+# import async_timeout
+import websockets
+
+
+def websocket_upgrade(http):
+    request_headers = dict(http.headers)
+    response_headers = []
+
+    def get_header(key):
+        key = key.lower().encode('utf-8')
+        return request_headers.get(key, b'').decode('utf-8')
+
+    def set_header(key, val):
+        response_headers.append((key.encode('utf-8'), val.encode('utf-8')))
+
+    try:
+        key = websockets.handshake.check_request(get_header)
+        websockets.handshake.build_response(set_header, key)
+    except websockets.InvalidHandshake:
+
+        # TODO: Handle invalid handshake
+
+        return
+
+    protocol = WebSocketProtocol(http, response_headers)
+    protocol.connection_open()
+    protocol.connection_made(http.transport, http.scope)
+    http.transport.set_protocol(protocol)
+
+
+async def reader(protocol):
+    close_code = None
+    order = 1
+    path = protocol.scope['path']
+    loop = protocol.loop
+    try:
+        async for data in protocol:
```

Comment:
Async iteration will not work with Python < 3.6, so I will change it back to using a loop/await when I have time tonight.